CDS/src/org/icpc/tools/cds/service/ContestRESTService.java

```diff
@@ -635,6 +641,149 @@ else if (!IContestObject.isSingleton(cType)) {
 		}
 	}
 
+	/**
+	 * Handle POST messages. The only thing you can currently post is a submission, either as a team
+	 * or as an admin user.
+	 */
+	@Override
+	protected void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+		if (!Role.isAdmin(request) && !Role.isTeam(request)) {
+			response.sendError(HttpServletResponse.SC_UNAUTHORIZED);
+			return;
+		}
+
+		String path = request.getPathInfo();
+		if (path == null || path.equals("/") || path.length() < 10) {
+			response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+			return;
+		}
+
+		String[] segments = path.substring(10).split("/");
+		if (segments == null || segments.length != 2) {
+			response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+			return;
+		}
+
+		ConfiguredContest cc = CDSConfig.getContest(segments[0]);
+		if (cc == null) {
+			response.sendError(HttpServletResponse.SC_NOT_FOUND, "Contest not found");
+			return;
+		}
+
+		String type = segments[1];
+		ContestType cType = IContestObject.getTypeByName(type);
```

Comment:
On phone so can’t verify but what happens if you supply an unknown type? Does it throw or return null?

Reply:
If the client tries to post to a valid contest but to anything other than /submission they receive an error 400 ("Invalid type"). I've added another commit to improve the message to "POST can only be used for submissions".